### PR TITLE
Add `generate_iodata` and performance improvements

### PR DESCRIPTION
**Pull request by jeroenvisser101**
This PR fixes a couple of smaller issues, along with introducing a new function that exposes the raw `iodata`, (or `chardata`, but I think how it's used in `xml_builder`, this is always `iodata`, as I believe all externally supplied content is `to_string`'d, but I may be wrong and  would be OK with renaming to `generate_chardata`, or altering this behaviour so it is guaranteed to output `iodata`).

Fixes:
- Uses `Enum.map_intersperse/3` (or at least, a copy of it, to keep support for Elixir versions before v1.10) instead of `Enum.map_join` or `Enum.map |> Enum.intersperse` uses.
- Silences warnings and asserts they are outputted when running tests (using ExUnit.CaptureIO)
- Replaces all string interpolation in favor of iodata/charlists
- Fixes CI by using `erlef/setup-beam@v1`, as `actions/setup-elixir` is deprecated and unmaintained.

### CODE REVIEW

**Comment by joshnuss** [comment]
Thanks @jeroenvisser101! Looks really good!

One thing I wasn't sure about, why we needed to replace `Enum.map(...) |> Enum.intersperse(...)` with `Enum.map_intersperse`?

**Comment by jeroenvisser101** [comment]
@joshnuss sorry! Meant to still have this as a draft, but appreciate the review already :)

`Enum.map(...) |> Enum.intersperse(...)` enumerates over the list twice, once to map it, and then another time to add the 'separator' after every element but the last. `map_intersperse` does this all at once, accumulating both the mapped element *and* the separator on each iteration but the first, where just the mapped element is added to the accumulator.

As the Elixir docs put more simply: "Maps and intersperses the given enumerable in one pass.".

I'm still tracking a few issues that have not been caught by the tests I added, and trying to see if I can assure the output is valid iodata (not just chardata), and that `iodata` is accepted as valid content (without requiring it to be converted to binary/bitstring).

I also want to look into the formatter and see if a backwards-compatible refactor could reduce the need to call it when the formatter is `none`, as now it still calls the function. We use `xml_builder` to generate very large feeds (>250M entries, each entry between 10-30 elements) and some of these elements are either the same for a few entries, the same for a part of the feed, or the same for all entries, and being able to generate them once and still keep them as `iodata` is likely to have quite a significant impact.

I'll update this PR over the next few days, and feel free to comment or give pointers.

Thanks 👍

**Comment by joshnuss** [comment]
Excellent, thanks for the explanation. One pass is definitely a good idea.

I think we should extract the `map_intersperse` to a separate module called `XmlBuilder.Compatibility`, that way we keep the top-level `XmlBuilder` module slimmer. Plus, it will also make it easier to replace with the `Enum.map_intersperse` when versions less than v1.10 become obsolete.

**Comment by devstopfix** [comment]
>  when versions less than v1.10 become obsolete.

I see 1.6 is already dropped - now work has begun on 1.13 - it could be a time to move up to 1.10 as a minimum? Then we drop OTP 20 as well and have OTP 21 as a minimum. See the [compatibility table](https://hexdocs.pm/elixir/1.12/compatibility-and-deprecations.html)

> I think we should extract the map_intersperse... 

I am not a licence expert - but an advantage of moving to 1.10 is that we would not have to redistribute Elixir code written under the Apache licence. My reading of this is that we would need to provide a notice and attribution of this: https://github.com/elixir-lang/elixir/blob/master/LICENSE#L89

**Comment by jeroenvisser101** [comment]
@devstopfix good call on licensing, if we keep this function we should add the license, and explicitly state copyright too.

Since xml_builder is so widely adopted, I'm not sure if dropping versions that are still supported by Elixir's core team would be a good idea, but I'm also not against it, since v1.10 has been out for 1.5 years by now.

There's also the option of, at compile-time, checking if `Enum.map_intersperse/3` is available, and falling back to the less optimal `Enum.map(...) |> Enum.intersperse(...)`, as users are most likely already on Elixir versions v1.10 or later, and accepting that older versions are slower? You can see an example in 1b10222. I think this implementation is also brief enough to not warrant another module, but can move if it you'd like @joshnuss.

PS. I found that [Floki does something similar](https://github.com/philss/floki/blob/e210d293fb08c8b7ec657983a836299dad9e9a05/lib/floki/raw_html.ex#L186), but doesn't specifically mention the license of `map_intersperse_list`. Since IANAL, I think opting for the suggestion in 1b10222 would be best, as there's no need to license anything, and it would run faster for those running newer Elixir versions and it would be equally fast for those who are not.

File: lib/xml_builder.ex
Comment (devstopfix) [inline lib/xml_builder.ex:425]: This is clever! I like this! 👍

**Comment by joshnuss** [comment]
@jeroenvisser101 the fallback option looks like the best way :+1:, nice work!

**Comment by jeroenvisser101** [comment]
@joshnuss @devstopfix I pushed 2 additional commits that add and change some functionality, namely:

- It now uses `IO.iodata_to_binary/1`, which in my benchmarks showed to be marginally faster than `chardata_to_string`. I added a `to_string` call on `encoding` to ensure that iodata is guaranteed to be valid, like we do elsewhere.
- Added 2 new ways to wrap content, `:safe` and `:iodata`. `:safe` skips any escaping, and we use it for types that we know to be numeric, and that would not ever return non-numeric data. This, in my tests, showed a real-world improvement of about 20% (we added it to about 50% of our fields). `:iodata`, which can appear both on the top level and within a list, will use the wrapped argument as-is, allowing you to generate static parts of the document beforehand. This is also documented (in addition to `generate_iodata` and the original escaping + `:cdata` type, with it's limitations and warnings outlined as well
- `quote_attribute_value` no longer contains a broken optimization where if the string would contain only a double or single quote, it would switch the quotes to the ones not used in the string. First of all, escaping was always done, so even if the value wouldn't require escaping, it would still escape it, and, secondly, the `single && double` check would never be required as the `escaped` data would be guaranteed not to contain either anymore. I've changed this behaviour to check only for 3 characters that should be encoded in attributes, `&`, `"` and `<`. If it contains neither of these, the string is safe and will not be escaped, and otherwise, will be escaped and always quoted in double quotes.
- I changed the macro's to be a bit more simple, I'm not sure if it has a runtime performance impact, but if it would, it's likely a positive one as there's less checks involved.

Let me know if you have any feedback/ideas about this, or if you'd want me to test/document/explain anything more.

Thanks 👍

**Comment by devstopfix** [comment]
>  which in my benchmarks showed to be marginally faster than 

I am curious how you make the benchmarks? I found this library superb! https://github.com/bencheeorg/benchee

**Comment by jeroenvisser101** [comment]
Yep, benchee with `Mix.install/2` on v1.12 :). Only thing that is tricky is to check 2 different versions of a package, but I usually test only a single function, so I copy it over to a dummy module to test.

The benchmarks of our real-world usage are based on throughput and timings, as well as `eprof` :)

**Comment by joshnuss** [comment]
@jeroenvisser101 LGTM! Let me know when you're ready for it to be merged, and I'll bump the version.

**Comment by jeroenvisser101** [comment]
Ready as can be :shipit:

**Comment by jeroenvisser101** [comment]
Let me rebase it before you do though!

**Comment by jeroenvisser101** [comment]
All good now, I think, if you can approve the Actions run, we'll be able to see if the actions run (like they do on my fork)

**Comment by jeroenvisser101** [comment]
Oh, I missed a formatting rule in tests.

File: lib/xml_builder.ex
Comment (joshnuss) [inline lib/xml_builder.ex:32]: Could this be shortened to:

```suggestion
    quote do: is_nil(unquote(map)) or unquote(map) == %{}
```

**Comment by joshnuss** [comment]
Really like the content wrapping `{:iodata,..}`, `{:safe, ..}`, very nice!

File: lib/xml_builder.ex
Comment (jeroenvisser101) [reply lib/xml_builder.ex]: Seems like we can! I thought maps always had to be escaped when used in macros, but turns out that's not true 👍

**Comment by joshnuss** [comment]
For the GitHub action change, not sure how to test those in a pull request.

Since it's broken on master anyways. I'm OK to merge and test it on master.

**Review by joshnuss — APPROVED** [review]

**Comment by joshnuss** [comment]
Published in [v2.2.0](https://hex.pm/packages/xml_builder/2.2.0)

**Comment by jeroenvisser101** [comment]
Awesome thanks! Also thank you @devstopfix for reviewing. Appreciate it!

**Comment by joshnuss** [comment]
Thanks for the contribution! Feel free to open PRs anytime.

**Pull request merged**
